---
title: "BCB420 Assignment 2"
date: '`r Sys.Date()`'
author: "Xiaolin Zhou"
output:
  html_document:
    toc: true
    toc_depth: 3
#bibliography: A2_Xiaolin_Zhou.bib
#nocite: '@*'
---

### Review of Assignment 1: data preparation and inital processing
<br>
**1. Install required packages in this assignment:**

```{r message=FALSE, results=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")
  }
if (!requireNamespace("GEOmetadb", quietly = TRUE)){
  BiocManager::install("GEOmetadb")
  }
library(GEOmetadb)

#install and load all packages need in A1
if (!requireNamespace('knitr', quietly = TRUE)){
install.packages('knitr', dependencies = TRUE)}
library(knitr)

if (!requireNamespace("magrittr", quietly = TRUE)){
install.packages("magrittr")}
library(magrittr) # needs to be run every time you start R and want to use %>%

if (!requireNamespace("kableExtra", quietly = TRUE)){
install.packages("kableExtra")}
library(kableExtra)

if (!requireNamespace("ggplot2", quietly = TRUE)){
install.packages("ggplot2")}
library(ggplot2)

BiocManager::install("edgeR")
library(edgeR)

```

**2. Download the selected data with supplementary files:**
* The expression data selected is _GSE160792_. And we initial processed the data by cleaning, mapping, and normalizing.
```{r message=FALSE, results=FALSE}
#get expression data
sfiles = getGEOSuppFiles('GSE160792')
 #<Detailed cleaning and mapping process is in "Assignment1.Rmd">
#We retreive the cleaned and mapped file for normalization
if (!file.exists("kas_filtered_mapped.rds")) {
  options(knitr.duplicate.label = 'allow')
  source(purl("Assignment1.Rmd", output = tempfile()))
}
kas_exp_filtered_annot <- readRDS("kas_filtered_mapped.rds")

```

**3. Applying TMM normalization to our dataset:**
```{r}
#Create an edgeR container for RNASeq count data
#use the filtered counts and make sure it is not a data.frame but is a matrix
filtered_data_matrix <- as.matrix(kas_exp_filtered_annot[,2:7]) 
rownames(filtered_data_matrix) <- kas_exp_filtered_annot$gene_id
#give the defined groups that we are going to be comparing - cell type (siMM treated or siRR treated)
d = DGEList(counts=filtered_data_matrix, group=c("Sample_K6_I", "Sample_K6_II", "Sample_K6_III","Sample_K1_I", "Sample_K1_II", "Sample_K1_III"))
#Calculate the normalization factors
d = calcNormFactors(d)
#get the normalized data
normalized_counts <- cpm(d)
```

**4. Preparing normalized counts data: **
<br>
* The normalized data with the desired format was saved into _GSE160792_finalized_normalized_counts_2022.txt_ file.
```{r message=FALSE, warning=FALSE, results='hide'}
#add the ensembl id column and HGNC symbol.
hgnc_col <- rownames(kas_exp_filtered_annot)
normalized_counts_data <- cbind(hgnc_col, normalized_counts)
ensembl_gene_id <- kas_exp_filtered_annot$ensembl_gene_id
normalized_counts_data <- cbind(ensembl_gene_id, normalized_counts_data)
normalized_counts_data <- as.data.frame(normalized_counts_data)

#remove duplicated ensembl ids
dup_id <- normalized_counts_data$ensembl_gene_id[duplicated(normalized_counts_data$ensembl_gene_id)]
normalized_counts_data <- normalized_counts_data[which(!normalized_counts_data$ensembl_gene_id %in% dup_id),]
#convert the elements in data frame to numeric value
i <- c(3, 4, 5, 6, 7, 8) 
normalized_counts_data[ , i] <- apply(normalized_counts_data[ , i], 2, function(x) as.numeric(as.character(x)))
if (!file.exists("GSE160792_finalized_normalized_counts_2022.txt")) {
write.table(normalized_counts_data, "GSE160792_finalized_normalized_counts_2022.txt", sep="\t")
}
```
---

### Differential Gene Expression:
<br>
**1. Install required packages in this assignment:**
```{r message=FALSE, warning=FALSE, results='hide'}
if (!requireNamespace("ComplexHeatmap", quietly = TRUE)){
  BiocManager::install("ComplexHeatmap")
}
library(ComplexHeatmap) 
library(circlize)

if (!require("kableExtra", quietly = TRUE)){
  install.packages("kableExtra")
}
library(kableExtra)

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("limma")
library(limma)

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

if (!requireNamespace("kableExtra", quietly = TRUE)){
install.packages("kableExtra")}
library(kableExtra)
```

**2. Load the data saved in Assignment 1:**
```{r message=FALSE, warning=FALSE}
normalized_counts_data <- read.table(file=file.path(getwd(),
                                   "GSE160792_finalized_normalized_counts_2022.txt"),
                                    header = TRUE,sep = "\t",
                                    stringsAsFactors = FALSE,
                                    check.names=FALSE)

kable(normalized_counts_data[1:5,1:5], type="html") 
```

**3. Visualize the data by heatmap:** 
```{r message=FALSE, warning=FALSE, fig.align='center', fig.dim=c(6,4)}
#Create a numerical matrix that we can create a heatmap from
heatmap_matrix <- normalized_counts_data[,3:ncol(normalized_counts_data)]
rownames(heatmap_matrix) <- normalized_counts_data$ensembl_gene_id
colnames(heatmap_matrix) <- colnames(normalized_counts_data[,3:ncol(normalized_counts_data)])

#create a heatmap
heatmap_matrix <- t(scale(t(heatmap_matrix)))
if(min(heatmap_matrix) == 0){
heatmap_col = colorRamp2(c( 0, max(heatmap_matrix)),
                         c( "white", "red"))
}else {
  heatmap_col = colorRamp2(c(min(heatmap_matrix), 0, 
                             max(heatmap_matrix)), c("blue", "white", "red"))}

current_heatmap <- Heatmap(as.matrix(heatmap_matrix),
      show_row_dend = TRUE,show_column_dend = TRUE,
      col=heatmap_col,show_column_names = TRUE,
      show_row_names = FALSE,show_heatmap_legend = TRUE)
#current_heatmap
```

**4. Check if AML differentially expressed in our samples:**
```{r message=FALSE, warning=FALSE}
#separate the positive and negative samples
kasumi_siRR_samples <- grep(colnames(normalized_counts_data),
                          pattern="?K6")
kasumi_siMM_samples <- grep(colnames(normalized_counts_data),
                          pattern="?K1")
gene_of_interest <- which(normalized_counts_data$hgnc_col == "RUNX1T1")

RUNX1T1_neg_samples <- t(normalized_counts_data[gene_of_interest, kasumi_siRR_samples])
colnames(RUNX1T1_neg_samples) <- c("RUNX1T1_knockdown_samples")
RUNX1T1_neg_samples

RUNX1T1_pos_samples <-t(normalized_counts_data[gene_of_interest, kasumi_siMM_samples])
colnames(RUNX1T1_pos_samples) <- c("RUNX1T1_samples")
RUNX1T1_pos_samples

#Is AML differentially expressed in our samples?
t.test(x=t(RUNX1T1_pos_samples),y=t(RUNX1T1_neg_samples))
```

**5. Use Limma to control for different variables:**
```{r message=FALSE, warning=FALSE}
#we hypothesize that RUNX1/RUNX1T1 knockdown will lead to the differential expression.

#Build the model.
#get the 2 and third token from the column names
samples <- data.frame( lapply(colnames(normalized_counts_data)[3:8], FUN=function(x){
          unlist(strsplit(x, split = "\\_"))[c(3,2)]}))
rownames(samples) <- c("replicates","cell_type")
colnames(samples) <- colnames(normalized_counts_data)[3:8]
samples <- data.frame(t(samples))
samples[1:5,]

#function to create a linear model in R - model.matrix 
#creates a design matrix
model_design <- model.matrix(~ samples$cell_type )
kable(model_design[1:5,], type="html")

#Create our data matrix
expressionMatrix <- as.matrix(normalized_counts_data[,3:8])
rownames(expressionMatrix) <- normalized_counts_data$ensembl_gene_id
colnames(expressionMatrix) <- colnames(normalized_counts_data)[3:8]
minimalSet <- ExpressionSet(assayData=expressionMatrix)
#Fit our data to the above model
fit <- limma::lmFit(minimalSet, model_design)

#Apply empircal Bayes to compute differential expression for the above described model. 
fit2 <- limma::eBayes(fit,trend=TRUE) #The parameter trend=TRUE is specific to RNA-seq data.
 
topfit <- topTable(fit2,
                   coef=ncol(model_design),
                   adjust.method = "BH",
                   number = nrow(expressionMatrix))
#merge hgnc names to topfit table
output_hits <- merge(normalized_counts_data[,1:2],
                     topfit,
                     by.y=0,by.x=1,
                     all.y=TRUE)
#sort by pvalue
output_hits <- output_hits[order(output_hits$P.Value),]
kable(output_hits[1:10,2:8],type="html",row.names = FALSE)

#How many gene pass the threshold p-value < 0.05?
length(which(output_hits$P.Value < 0.05)) #5907

#How many genes pass correction?
length(which(output_hits$adj.P.Val < 0.05)) #3624
```

**6. Improve our results if we account for the replicate variability:**
```{r message=FALSE, warning=FALSE}
#function to create a linear model in R - model.matrix creates a design matrix
model_design_pat <- model.matrix(~ samples$replicates + samples$cell_type)
kable(model_design_pat,type="html")

#Fit our data to the above model
fit_pat <- lmFit(minimalSet, model_design_pat)
#Apply empircal Bayes to compute differential expression for the above described model.
fit2_pat <- eBayes(fit_pat,trend=TRUE)
 
topfit_pat <- topTable(fit2_pat,
                   coef=ncol(model_design_pat),
                   adjust.method = "BH",  # we use Benjamni - hochberg for hypothesis correction
                   number = nrow(expressionMatrix))
#merge hgnc names to topfit table
output_hits_pat <- merge(normalized_counts_data[,1:2],
                         topfit_pat,by.y=0,by.x=1,all.y=TRUE)
#sort by pvalue
output_hits_pat <- output_hits_pat[order(output_hits_pat$P.Value),]
kable(output_hits_pat[1:10,2:8],type="html",row.names = FALSE)

#How many gene pass the threshold p-value < 0.05?
length(which(output_hits_pat$P.Value < 0.05)) # [1] 5761
#How many genes pass correction?
length(which(output_hits_pat$adj.P.Val < 0.05)) # [1] 3464
```

**7. Use edgeR package for processing RNASeq data and calculating the differential expression:**
```{r message=FALSE, warning=FALSE, fig.align='center'}
#Set up our edgeR objects
d = DGEList(counts=filtered_data_matrix, group=samples$cell_type)
#Estimate Dispersion - our model design.
d <- estimateDisp(d, model_design_pat)
#Fit the model
fit <- glmQLFit(d, model_design_pat)
#kable(model_design_pat[1:3,1:3], type="html") %>% row_spec(0, angle = -45)
#Calculate differential expression using the Quasi liklihood model
qlf.pos_vs_neg <- glmQLFTest(fit, coef='samples$cell_typeK6')
kable(topTags(qlf.pos_vs_neg), type="html",row.names = FALSE)

#Get all the results
qlf_output_hits <- topTags(qlf.pos_vs_neg,sort.by = "PValue",
                           n = nrow(normalized_counts_data))
rownames(qlf_output_hits$table) <- normalized_counts_data$ensembl_gene_id
#How many gene pass the threshold p-value < 0.05?
length(which(qlf_output_hits$table$PValue < 0.05)) #6072
#How many genes pass correction?
length(which(qlf_output_hits$table$FDR < 0.05)) #4276
```

* Compare the results from the two different models Limma vs Quasi liklihood
```{r message=FALSE, warning=FALSE, fig.align='center', echo=FALSE, out.width=c('50%', '50%'), fig.show='hold', fig.cap="Comparison of results from limma and edgeR models"}
qlf_pat_model_pvalues <- data.frame(ensembl_id = rownames(qlf_output_hits$table),
                                    qlf_patient_pvalue=qlf_output_hits$table$PValue)
limma_pat_model_pvalues <- data.frame(ensembl_id = output_hits_pat$ensembl_gene_id,
                                      limma_patient_pvalue = output_hits_pat$P.Value)
two_models_pvalues <- merge(qlf_pat_model_pvalues,
                            limma_pat_model_pvalues,
                            by.x=1,by.y=1)
two_models_pvalues$colour <- "grey"
two_models_pvalues$colour[two_models_pvalues$qlf_patient_pvalue
                          <0.05] <- "orange"
two_models_pvalues$colour[two_models_pvalues$limma_patient_pvalue
                          <0.05] <- "blue"
two_models_pvalues$colour[two_models_pvalues$qlf_patient_pvalue
                          <0.05 &
two_models_pvalues$limma_patient_pvalue<0.05] <- "red"

plot(two_models_pvalues$qlf_patient_pvalue,
     two_models_pvalues$limma_patient_pvalue,
     col = two_models_pvalues$colour,
     xlab = "QLF patient model p-values",
     ylab ="Limma Patient model p-values",
     main="QLF vs Limma")

legend(0,1,legend=c("qlf","limma","both","not sig"),
       fill=c("orange","blue","red","grey"),cex = 0.7)

#highlight the gene of interest
ensembl_of_interest <- normalized_counts_data$ensembl_gene_id[
  which(normalized_counts_data$hgnc_col == "RUNX1T1")]
two_models_pvalues$colour <- "grey"
two_models_pvalues$colour[two_models_pvalues$ensembl_id==ensembl_of_interest] <- "red"
plot(two_models_pvalues$qlf_patient_pvalue,
     two_models_pvalues$limma_patient_pvalue,
     col = two_models_pvalues$colour,
     xlab = "QLF patient model p-values",
     ylab ="Limma Patient model p-values",
     main="QLF vs Limma")
points(two_models_pvalues[
  two_models_pvalues$ensembl_id==ensembl_of_interest,2:3],
       pch=24,  col="red", cex=1.5)
```





**7.Compare the results from the two different models and highlight the gene of interest:**
```{r message=FALSE, warning=FALSE, fig.align='center'}
simple_model_pvalues <- data.frame(ensembl_id = output_hits$ensembl_gene_id,
                                   simple_pvalue=output_hits$P.Value)
MHT_model_pvalues <-  data.frame(ensembl_id = output_hits_pat$ensembl_gene_id,
                                 rep_pvalue = output_hits_pat$P.Value)
two_models_pvalues <- merge(simple_model_pvalues, MHT_model_pvalues,by.x=1,by.y=1)
two_models_pvalues$colour <- "black"
two_models_pvalues$colour[two_models_pvalues$simple_pvalue<0.05] <- "orange" #genes sig in simple model
two_models_pvalues$colour[two_models_pvalues$rep_pvalue<0.05] <- "blue"  #genes sig in rep model with BH multiple hypothesis
two_models_pvalues$colour[two_models_pvalues$simple_pvalue<0.05 & two_models_pvalues$rep_pvalue<0.05] <- "red"
#genes sig in both models

#visualization for sig genes
plot(two_models_pvalues$simple_pvalue, two_models_pvalues$rep_pvalue,
     col = two_models_pvalues$colour,
     xlab = "Simple model p-values",
     ylab ="Multiple hypothesis model p-values",
     main="Simple vs Replicate Limma")

#Visualize the sig value for RUNX1T1
ensembl_of_interest <- normalized_counts_data$ensembl_gene_id[
  which(normalized_counts_data$hgnc_col == "RUNX1T1")]
two_models_pvalues$colour <- "grey"
two_models_pvalues$colour[two_models_pvalues$ensembl_id==ensembl_of_interest] <- "red"

plot(two_models_pvalues$simple_pvalue,
     two_models_pvalues$rep_pvalue,
     col = two_models_pvalues$colour,
     xlab = "Simple model p-values",
     ylab ="Replicate model p-values",
      main="Simple vs Replicate Limma")

points(two_models_pvalues[which(
  two_models_pvalues$ensembl_id == ensembl_of_interest),2:3],
       pch=20, col="red", cex=1.5)
legend(0,1,legend=c("RUNX1T1","rest"),
       fill=c("red","grey"),cex = 0.7)

```

**8. Draw volcano plot or MA plot with interested gene highlighted:**
<br>

* a) Volcano plot for amount of differentially expressed genes in **simple** model and highlight the genes of interest.
```{r message=FALSE, warning=FALSE, fig.align='center', echo=FALSE, out.width=c('50%', '50%'), fig.show='hold', fig.cap="Figure: Kasumi-1 AML cells Volcano Plots for Simple and Replicate models", fig.align='center'}
#To better visualize the graph, we first filtered out the genes with extremely logFC.
filtered_output_hits <- output_hits[which(output_hits$logFC < 150 & output_hits$logFC > -150),]
ensembl_of_interest <- normalized_counts_data$ensembl_gene_id[
  which(normalized_counts_data$hgnc_col == "RUNX1T1")]

df_limma <- data.frame(ensembl_gene_id = filtered_output_hits$ensembl_gene_id,
                 logFC = filtered_output_hits$logFC,
                 logpvalue = -log10(filtered_output_hits$P.Value),
                 adj_p = filtered_output_hits$adj.P.Val,
                 color="grey")
df_limma$color[which(df_limma$logFC > 0 & df_limma$adj_p < 0.05)] <- "green"
df_limma$color[which(df_limma$logFC< 0 & df_limma$adj_p < 0.05)] <- "blue"
df_limma$color[df_limma$ensembl_gene_id==ensembl_of_interest] <- "red"
p_limma <- plot(df_limma[,c(2,3)], col=df_limma$color, main = "Volcano plot for P-value and LogFoldChange Calculated by Limma method")

points(filtered_output_hits[which(
  filtered_output_hits$ensembl_gene_id == ensembl_of_interest),2:3],
       pch=20, col="red", cex=1.5)
legend("topright", 
       legend=c("up-regulated", "down-regulated", "rest", "gene of interest"), 
       fill = c("blue","green", "grey", "red"),
       cex = 0.7)
```
* b) Volcano plot for amount of differentially expressed genes in **replicate** model built by edgeR and highlight the genes of interest.
```{r message=FALSE, warning=FALSE, fig.align='center', echo=FALSE, out.width=c('50%', '50%'), fig.show='hold', fig.cap="Figure: Kasumi-1 AML cells Volcano Plots for Simple and Replicate models", fig.align='center'}
#To better visualize the graph, we first filtered out the genes with extremely logFC.
filtered_output_hits_pat <- output_hits_pat[which(output_hits_pat$logFC < 150 & output_hits_pat$logFC > -150),]
ensembl_of_interest <- normalized_counts_data$ensembl_gene_id[
  which(normalized_counts_data$hgnc_col == "RUNX1T1")]

df_edger <- data.frame(ensembl_gene_id = filtered_output_hits_pat$ensembl_gene_id,
                 logFC = filtered_output_hits_pat$logFC,
                 logpvalue = -log10(filtered_output_hits_pat$P.Value),
                 adj_p = filtered_output_hits_pat$adj.P.Val,
                 color="grey")
df_edger$color[which(df_edger$logFC > 0 & df_edger$adj_p < 0.05)] <- "green"
df_edger$color[which(df_edger$logFC< 0 & df_edger$adj_p < 0.05)] <- "blue"
df_edger$color[df_edger$ensembl_gene_id==ensembl_of_interest] <- "red"
p_edger <- plot(df_edger[,c(2,3)], col=df_edger$color, main = "Volcano plot for P-value and LogFoldChange Calculated by edgeR method")

points(filtered_output_hits_pat[which(
  filtered_output_hits_pat$ensembl_gene_id == ensembl_of_interest),2:3],
       pch=20, col="red", cex=1.5)
legend("topright", 
       legend=c("up-regulated", "down-regulated", "rest", "gene of interest"), 
       fill = c("blue","green", "grey", "red"),
       cex = 0.7)
```
<br>

**9.Compare the results from the two different models and highlight the gene of interest:**
* Visualize the top hits use a heatmap
```{r message=FALSE, warning=FALSE, fig.align='center', fig.cap="Heatmap of Tophits (p-value<0.05) of Gene expressions", , fig.dim=c(6,4)}
top_hits <- output_hits_pat$ensembl_gene_id[
  output_hits_pat$P.Value<0.05]
heatmap_matrix_tophits <- t(
  scale(t(heatmap_matrix[
which(rownames(heatmap_matrix) %in% top_hits),])))
if(min(heatmap_matrix_tophits) == 0){
heatmap_col = colorRamp2(c( 0, max(heatmap_matrix_tophits)),
                             c( "white", "red"))
} else {
heatmap_col = colorRamp2(c(min(heatmap_matrix_tophits), 0,
      max(heatmap_matrix_tophits)), c("blue", "white", "red"))
  }
tophits_heatmap <- Heatmap(as.matrix(heatmap_matrix_tophits),
                           cluster_rows = TRUE,
                           cluster_columns = TRUE,
                               show_row_dend = TRUE,
                               show_column_dend = TRUE,
                               col=heatmap_col,
                               show_column_names = TRUE,
                               show_row_names = FALSE,
                               show_heatmap_legend = TRUE,
                               )
tophits_heatmap
```

* Order the heatmap by knock-off genes
```{r message=FALSE, warning=FALSE, fig.align='center', fig.cap="Heatmap of Tophits (p-value<0.05) of Gene expressions ordered by knockdown genes", fig.dim=c(6,4)}
heatmap_matrix_tophits <- heatmap_matrix_tophits[,
c(grep(colnames(heatmap_matrix_tophits),
                      pattern = "?K6"),
  grep(colnames(heatmap_matrix_tophits),
                      pattern = "?K1"))]
if(min(heatmap_matrix_tophits) == 0){
heatmap_col = colorRamp2(c( 0, max(heatmap_matrix_tophits)),
                             c( "white", "red"))
} else {
heatmap_col = colorRamp2(c(min(heatmap_matrix_tophits), 0,
                    max(heatmap_matrix_tophits)),
                    c("blue", "white", "red"))
}
ordered_tophits_heatmap <- Heatmap(as.matrix(heatmap_matrix_tophits),
                           cluster_rows = TRUE,
                           cluster_columns = FALSE, #do not cluster the columns
                               show_row_dend = TRUE,
                               show_column_dend = TRUE,
                               col=heatmap_col,
                               show_column_names = TRUE,
                               show_row_names = FALSE,
                               show_heatmap_legend = TRUE,
                               )
ordered_tophits_heatmap 
```


* Set p_value threshold < 0.01 to get more cleaner picture
```{r message=FALSE, warning=FALSE, fig.align='center', fig.cap="Heatmap of Tophits (p-value<0.01) of Gene expressions ordered by knockdown genes"}
top_hits <- output_hits_pat$ensembl_gene_id[output_hits_pat$P.Value<0.01]
heatmap_matrix_tophits <- t(
scale(t(heatmap_matrix[which(rownames(heatmap_matrix) %in% top_hits),])))
heatmap_matrix_tophits<- heatmap_matrix_tophits[,
       c(grep(colnames(heatmap_matrix_tophits),pattern = "?K6"),
         grep(colnames(heatmap_matrix_tophits),pattern = "?K1"))]
if(min(heatmap_matrix_tophits) == 0){
heatmap_col = colorRamp2(c( 0, max(heatmap_matrix_tophits)),
                             c( "white", "red"))
} else {
heatmap_col = colorRamp2(c(min(heatmap_matrix_tophits), 0,
                              max(heatmap_matrix_tophits)),
                             c("blue", "white", "red"))
}
ordered_restricted_tophits_heatmap <- Heatmap(as.matrix(heatmap_matrix_tophits),
                cluster_rows = TRUE,  show_row_dend = TRUE,
                cluster_columns = FALSE,show_column_dend = FALSE,
                col=heatmap_col,show_column_names = TRUE,
                show_row_names = FALSE,show_heatmap_legend = TRUE)

ordered_restricted_tophits_heatmap

```


### Summary for differential gene expression analysis:
1. **Calculate p-values for each of the genes in your expression set.** How many genes were significantly differentially expressed? What thresholds did you use and why?
  - We constructed two models by limma and edgeR pacakges to calculate the p-values for each of the genes. The threshold was set to **0.05**, which is the common threshold for p-value to indicate the significance of the expression data. P-value < 0.05 represents that the possibility to reject the null hypothesis is more than 95%.
  - For the model built by limma, there were **5907** genes that were significantly differentitally expressed with the threshold of 0.05. Also, **3624** genes passed the correction.
  - For the model built by edgeR, there were **5761** genes that were significantly differentitally expressed with the threshold of 0.05. Also, **3464** genes passed the correction.

2. **Multiple hypothesis testing - correct your p-values using a multiple hypothesis correction method.** Which method did you use? And Why? How many genes passed correction?
  -  I used **Benjamini-Hochberg** (BH) method to correct the p-value since it is a common method used in multiple hypothesis testing. BH is not overly stringent and can be used to control for the family-wise error rate and false discovery rate.
  - For the model built by limma, there were **3624** genes passed the correction.
  - For the model built by edgeR, there were **3464** genes passed the correction.
  
3. **Show the amount of differentially expressed genes using an MA Plot or a Volcano plot. Highlight genes of interest.**
  - As the heatmap shown above, we separated the genes by up-regulated and down-regulated categories. There are clear divisons on the heatmap. The left region indicates the expression of siRR-treated replicates with RUNX1/RUNXT1 knocked-off, while the right region indicates the expression of siMM-treated replicates with RUNX1/RUNXT1 gene. We could clearly observed that expressions for up-regulated and down-regulated genes were separated by color. And the differential expression data processed by edgeR package offered a more clear division. 

4. **Visualize your top hits using a heatmap.** Do you conditions cluster together? Explain why or why not.
  - Yes, the results from limma and edgeR showed clusters of siRR-treated samples and siMM-treated samples, indicating that RUNX1/RUNXT1 plays an essential role in regulating the cell physiological process.






### Threshold over-representation analysis:
```{r message=FALSE, warning=FALSE}
#How many genes are up regulated?
length(which(qlf_output_hits$table$PValue < 0.05 & qlf_output_hits$table$logFC > 0))
## [1] 2727
#How many genes are down regulated?
length(which(qlf_output_hits$table$PValue < 0.05 & qlf_output_hits$table$logFC < 0))
## [1] 3345

#Create thresholded lists of genes.
#merge gene names with the top hits
qlf_output_hits_withgn <- merge(normalized_counts_data[,1:2], qlf_output_hits, by.x=1, by.y=0)

qlf_output_hits_withgn[,"rank"] <- -log(qlf_output_hits_withgn$PValue, base=10)*sign(qlf_output_hits_withgn$logFC)
qlf_output_hits_withgn <- qlf_output_hits_withgn[order(qlf_output_hits_withgn$rank),]

upregulated_genes <- qlf_output_hits_withgn$hgnc_col[
  which(qlf_output_hits_withgn$PValue < 0.05 & qlf_output_hits_withgn$logFC > 0)]
downregulated_genes <- qlf_output_hits_withgn$hgnc_col[
  which(qlf_output_hits_withgn$PValue < 0.05 & qlf_output_hits_withgn$logFC < 0)]

write.table(x=upregulated_genes,
            file=file.path("~/projects/Desktop/Xiaolin/data","AML_upregulated_genes.txt"),sep="\t",
            row.names = FALSE,col.names = FALSE,quote = FALSE)
write.table(x=downregulated_genes,
            file=file.path("data","AML_downregulated_genes.txt"),sep="\t",
            row.names = FALSE,col.names = FALSE,quote = FALSE)
write.table(x=data.frame(genename= qlf_output_hits_withgn$hgnc_col,F_stat=qlf_output_hits_withgn$rank),
            file=file.path("data","AML_ranked_genelist.txt"),sep = "\t",
            row.names = FALSE,col.names = FALSE,quote = FALSE)

sig_differential_expressed <- qlf_output_hits_withgn$hgnc_col[
  which(qlf_output_hits_withgn$PValue < 0.05)]  
write.table(x=sig_differential_expressed,
            file=file.path("data","AML_complete_genelist.txt"),sep = "\t",
            row.names = FALSE,col.names = FALSE,quote = FALSE)
            
```
* Questions for over-representation analysis
1. Which method did you choose and why?
   - I chose g:profiler since it is a well-known web server for gene enrichment analysis and it includes more than 9 annotation gene sets, such as GO and KEGG.
   
2. What annotation data did you use and why? What version of the annotation are you using?
   - I used GO, Reactome, and Wikipathways since they contain comprehensive functional annotations of human genes. GO covers 43,786 GO terms, 7,965,896 annotations, and 1,566,018 gene products. The g:profiler involves the GO released on 2022-01-13, version 79 of Reactome, and Wikipathway released on 2022.
  
3. How many genesets were returned with what thresholds?
   - With the threshold set to 0.05, we obtained three genesets --- GO biological process (Go: BP), Reactome (REAC), and Wikipathways (WP).
   * Results from up_regulated genes list:
```{r results='asis', echo=FALSE}
#The top term returned in each data source:
data_resource = c("Go: BP", "REAC", "WP")
top_terms = c("cellular macromolecule metabolic process", 
              "RHO GTPase cycle", 
              "TNF-alpha signaling pathway")
t = c(5746,444,94)
q = c(2544,1718,1247)
t_intersect_q = c(1056, 107, 32)
df_up = data.frame(data_resource, top_terms, t, q, t_intersect_q)
colnames(df_up) = c("Data source", "Top Term", "T", "Q", "T intersect Q")
kable(df_up, format = 'html',  escape=FALSE,align=c(rep('c',times=5)))%>%
  kable_styling(full_width = F)
```
<br>

   * Results from down_regulated genes list:
```{r results='asis', echo=FALSE}
#The top term returned in each data source:
data_resource = c("Go: BP", "REAC", "WP")
top_terms = c("cellular macromolecule metabolic process", 
              "Metabolism of RNA", 
              "DNA repair pathways, full network")
t = c(5746,661,120)
q = c(3144,2126,1524)
t_intersect_q = c(1269, 195, 46)
df_down = data.frame(data_resource, top_terms, t, q, t_intersect_q)
colnames(df_down) = c("Data source", "Top Term", "T", "Q", "T intersect Q")
kable(df_down, format = 'html',  escape=FALSE,align=c(rep('c',times=5)))%>%
  kable_styling(full_width = F)
```
<br>

* Results for all differentially expressed genes:
```{r go_profiler_all, results='asis', echo=FALSE}
#The top term returned in each data source:
data_resource = c("Go: BP", "REAC", "WP")
top_terms = c("cellular macromolecule metabolic process", 
              "Metabolism of RNA", 
              "Cell cycle")
t = c(5746,661,122)
q = c(5694,3848,2772)
t_intersect_q = c(2327, 345, 75)
df_all = data.frame(data_resource, top_terms, t, q, t_intersect_q)
colnames(df_all) = c("Data source", "Top Term", "T", "Q", "T intersect Q")
kable(df_all, format = 'html',  escape=FALSE,align=c(rep('c',times=5)))%>%
  kable_styling(full_width = F)
```
<br>


4. Run the analysis using the up-regulated set of genes, and the down-regulated set of genes separately. How do these results compare to using the whole list (i.e all differentially expressed genes together vs. the up-regulated and down regulated differentially expressed genes separately)?
 - The separate results are shown above. From the tables we could observe that the up-regulated, down-regulated genes, and the complete list of genes have the same top term (cellular macromolecule metabolic process) for GO biological process. However, the top terms from Reactome and Wikipathways are different.
 - The whole list results are more close to the down_regulated results. From the difference of up-regulated genes and complete gene list, we could observe that `r df_all[1,5]-df_up[1,5]` of down_regulated genes were up-regulated in the whole list in GO biological pathway. Moreover, `r df_all[2,5]-df_up[2,5]` of down_regulated genes were up-regulated in the whole list in Reactome. In addition, `r df_all[3,5]-df_up[3,5]` of down_regulated genes were up-regulated in the whole list in Wikipathways.

---

### Interpretation:
1. Do the over-representation results support conclusions or mechanism discussed in the original paper?
* Yes, the majority of the over-representation results (ORA) support conclusions discussed in the paper. In the original paper, the gene set enrichment analysis suggests that knockdown of RUNX1/RUNX1T1 influences RNA binding proteins as well as snRNP assembly, and is associated with impaired mRNA processing, especailly the splicing pathways. Our over-representation results indicated that the pathways or chemical reactions involve macromolecules with relatively high molecular mass might be affected since the top term from functionality analysis for up-regulated, down-regulated and complete list of genes is **cellular macromolecule metabolic process**. The **snRNP** and **RNA binding proteins** are considered as macromolecules; therefore, the ORA results in accord with the mechanisms mentioned in the paper. 
* Moreover, the paper also cluster the gene sets with similar functions. The third largest cluster is related to **mRNA/rRNA metabolic process**, which supports our ORA results from down-regulated genes and complete gene list since the top term returned by Reactome database is **Metabolism of RNA**.

2. Can you find evidence, i.e. publications, to support some of the results that you see. How does this evidence support your results.
* RUNX1/RUNX1T1 binding sites are often positioned near genes encoding splicing-associated factors or in their immediate vicinity (Grinev et al., 2019). Such genes includes classical splicing regulatory genes such as **HNRNPM** and **RBFOX2**. HNRNPM could be found in differential expressed down-regulated genes, while RBFOX2 presented in differential expressed up-regulated genes.
* Additionally, STUB1 downregulated RUNX1– RUNX1T1 expression and induced ubiquitination of the enzyme. Interestingly, STUB1 overexpression inhibited leukemia cells harboring RUNX1– RUNX1T1, but was ineffective in non-R-expressing cells (Yonezawa et al., 2017). Our results show that STUB1 was significantly expressed in over-representation analysis. And STUB1 bound to RUNX1 and induced its ubiquitination and resulted in degradation mainly in the nucleus (Yonezawa et al., 2017), which explains the top term of **Cell cycle** returned by complete gene list.

### Additional Links:

* The link of journal: 

  - [A2 journal](https://github.com/bcb420-2022/Xiaolin_Zhou/wiki/BCB420-Assignment-2:-Differential-Gene-expression-and-Preliminary-ORA)

* Screenshots for g:Profiler:

  - [up-regulated overview](data/gProfiler_upregulated.png)
  - [up-regulated details](data/gProfiler_upregulated_details.png)
  - [down-regulated overview](data/gProfiler_downregulated.png)
  - [down-regulated details](data/gProfiler_downregulated_details.png)
  - [complete list overview](data/gProfiler_complete_gene_list.png)
  - [complete list details](data/gProfiler_complete_gene_list_details.png)

---

### Reference:

1. Isserlin, R. (2021). Lecture 6: Differential Expression. BCB420 - Computational System Biology. Retrieved March, 10, 2022

2. Isserlin, R. (2021). Lecture 7: Annotation Dataset and Intro to Pathway Analysis. BCB420 - Computational System Biology. Retrieved March, 10, 2022.

3. Grinev, V.V., Barneh, F., Ilyushonak, I. M., Nakjang, S., Smink, J., van Oort, A., Clough, R., Seyani, M., McNeill, H., Reza, M., Martinez-Soria, N., Assi, S. A., Ramanouskaya, T. V., Bonifer, C., &amp; Heidenreich, O. (2021). Runx1/RUNX1T1 mediates alternative splicing and reorganises the transcriptional landscape in leukemia. *Nature News*. Retrieved February 19, 2022, from https://www.nature.com/articles/s41467-020-20848-z

4. Grinev, V., Ilyushonak, I., Clough, R., Nakjang, S., Smink, J., Martinez-Soria, N., Ramanouskaya, T., Bonifer, C., &amp; Heidenreich, O. (2019, January 1). Runx1/RUNX1T1 controls alternative splicing in the t(8;21)-positive acute myeloid leukemia cells. *bioRxiv*. Retrieved March 15, 2022, from https://doi.org/10.1101/628040 

5. Yonezawa T., Takahashi H., Shikata S., Liu X., Tamura M., Asada S., Fukushima T., Fukuyama T., Tanaka Y., Sawasaki T., Kitamura T., Goyama S., (2017). The ubiquitin ligase stub1 regulates stability and activity of RUNX1 and runx1-RUNX1T1. The Journal of biological chemistry. Retrieved March 15, 2022, from https://pubmed.ncbi.nlm.nih.gov/28536267/ 

